Add README

olafurpg opened the pull request:
- Remove Enumeratum dependency in jsonrpc module. It was only
  used for ErrorCode and then it turned out that the JSON
  decoder for `ErrorCode` was buggy since it would fail on
  unknown error codes.
- Document dependencies
- Installation instructions
- Gettings started instructions
- Document alternatives

Fixes #8

olafurpg committed:
Add README

- Remove Enumeratum dependency in jsonrpc module. It was only
  used for ErrorCode and then it turned out that the JSON
  decoder for `ErrorCode` was buggy since it would fail on
  unknown error codes.
- Document dependencies
- Installation instructions
- Gettings started instructions
- Document alternatives

olafurpg commented:
PS @jvican and @jastice, we should update the `bsp` module to depend on `jsonrpc` instead of `lsp4s`. That should shave off around 1.5mb, in combination with removing shapeless in #9, pprint in #10 and enumeratum in this PR the next bsp release could be ~5mb smaller.

gabro commented on README.md:
```diff
@@ -1,13 +1,213 @@
+
+[![Maven Central](https://maven-badges.herokuapp.com/maven-central/org.scalameta/lsp4s_2.12/badge.svg)](https://maven-badges.herokuapp.com/maven-central/org.scalameta/lsp4s_2.12)
+
+To implement a language server, add the following dependency you need.
```
This sounds like Yoda 😅

olafurpg replied:
Haha, good catch, was a bad rephrase from an old text

gabro reviewed (approved):
👌 two minor comments.
Thanks for writing this, it's extremely useful!

olafurpg committed:
Address review feedback

olafurpg committed:
Minor polish

olafurpg committed:
Disable parallel execution

olafurpg commented:
Weird, the tests started failing consistently on 2.12 only and it seems unrelated to the PR 🤔 will need to investigate...

jvican commented:
I’ll try to review tonight, if I’m blocking something feel free to merge!

olafurpg commented:
There's no rush @jvican

olafurpg committed:
Re-enable parallel execution

olafurpg committed:
Run 2.11 and 2.12 together

olafurpg committed:
Ensure that 2 pongs have returned before running assertion

olafurpg committed:
Refactor ping pong suite to be clearer and more robust

- rename "client" and "server" into "Alice" and "Bob"
- rename "ClientServer" into "TestConnection"
- use Promise.complete to ensure pong responses have been received
- cancel input/output streams for ClientServer

olafurpg commented:
My conclusion is that the ping pong suite was flaky and it just happened to pass the CI in #10. The tests failed in two different ways

- Timeout, all tests had passed but the sbt shell never exited

```
[info] Passed: Total 21, Failed 0, Errors 0, Passed 21
[info] Done compiling.
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```

- Assertion error: the second pong response never delivered

```
[info] - ping pong *** FAILED ***
[info]   received List(Pong from client) != expected List(Pong from client, Pong from server) (PingPongSuite.scala:68)
```

I've refactored in the ping pong suite to

- ensure that we close all input/output streams from the test connection. This should fix the timeout error.
- ensure that the assertions run after all pongs have been received by using `Promise.complete`, previously we relied on  the the scheduling from the execution context. This should fix the assertion error.

While refactoring, I found it awkward to read `connection.client.client`/`connection.server.client`. Given that JSON-RPC is bi-directional I renamed `connection.client` to `connection.alice` and `connection.server` to `connection.bob`.

I've restarted the CI build three times now and the tests have always passed.

olafurpg committed:
Minor updates to prose

olafurpg commented:
Merging to fix flaky tests on master. Feel free to follow up later once you have time to review @jvican
